### PR TITLE
Merge with upstream

### PR DESCRIPTION
**Pull request by wes-turner**
Merges upstream changes on top of our master.  With as many commits as we had, it was too much work to rebase our own changes to atop upstream.

### CODE REVIEW

**Comment by wes-turner** [comment]
https://github.com/google/dopamine/compare/master..wes-turner:merge-with-upstream (double-dot compare rather than GitHub's default triple-dot) to compare this branch with upstream master.

**Comment by wes-turner** [comment]
After the fix from PR #23, `./tests/dopamine/agents/implicit_quantile/implicit_quantile_agent_test.py` is the only remaining failing test file.

**Comment by wes-turner** [comment]
I take it back.  `./tests/dopamine/agents/implicit_quantile/implicit_quantile_agent_test.py` is broken *in* master, but is not in this branch (either with or without the #23 patch).  `./tests/dopamine/tests/threading_utils_integration_test.py` is now broken as a result of the merge (something changed out from under it).  I'll fix that in a few minutes.  All other tests now pass.

I'd also thought there was some silent breakage from reading through test logs for assertions that had failed but were not somehow counting as failed tests.  Those turn out to be from tests that seek to verify assertions (like `testProjectNewSupportHasInconsistentDeltask` testing https://github.com/wes-turner/dopamine/blob/master/dopamine/agents/rainbow/rainbow_agent.py#L388).  I think, though, that out of general frustration I've brought on myself from trying to juggle too many things, I'd like to comment that this behavior is non-obvious from the test name and that the style guide forbids using `assert` in this way:

> Do not use assert statements for validating argument values of a public API. assert is used to ensure internal correctness, not to enforce correct usage

**Comment by wes-turner** [comment]
Re: `tf.Assert`, could it be that it's a necessary evil because one needs evaluate the tensors within a session?  This seems like something I should know the answer to (and don't).

**Comment by wes-turner** [comment]
Fixed with [latest commit](https://github.com/wes-turner/dopamine/pull/24/commits/41274215b8eab297a259160d60d95e9c39ff82f8).  Nice work, Arthur, catching pre-emptively with your new tests.

All tests now pass.

File: dopamine/agents/dqn/dqn_agent.py [hunk 1]
```diff
@@ -528,6 +527,9 @@ def unbundle(self, checkpoint_dir, iteration_number, bundle_dictionary):
     except tf.errors.NotFoundError:
       return False
     for key in bundle_dictionary:
+      # Since some of self's attributes are emulated with
```
Comment (arthurarg) [inline dopamine/agents/dqn/dqn_agent.py:530]: Thanks for adding this, it's helpful.

**Review by arthurarg — APPROVED** [review]

**Comment by arthurarg** [comment]
Sorry for the delay, LGTM

**Comment by wes-turner** [comment]
No worries about the delay.  Thanks for reviewing it.

**Pull request merged**